rustls/src/crypto/aws_lc_rs/hpke.rs

```diff
@@ -0,0 +1,602 @@
+use alloc::sync::Arc;
+use alloc::vec;
+use alloc::vec::Vec;
+use core::fmt::{self, Debug, Formatter};
+
+use aws_lc_rs::aead::{
+    Aad, BoundKey, Nonce as AwsNonce, NonceSequence, OpeningKey, SealingKey, UnboundKey,
+};
+use aws_lc_rs::hkdf::KeyType;
+use aws_lc_rs::{aead, agreement, hkdf, hmac};
+
+use crate::crypto::aws_lc_rs::tls13::RingHkdf;
+use crate::crypto::hpke::{
+    EncapsulatedSecret, Hpke, HpkeOpener, HpkePrivateKey, HpkeProvider, HpkePublicKey, HpkeSealer,
+    HpkeSuite,
+};
+use crate::crypto::tls13::{Hkdf, HkdfExpander, OutputLengthError};
+use crate::msgs::enums::{HpkeAead, HpkeKdf, HpkeKem};
+use crate::{Error, OtherError};
+
+/// An [RFC 9180] Hybrid Public Key Encryption (HPKE) provider backed by aws-lc-rs cryptography.
+///
+/// [RFC 9180]: https://www.rfc-editor.org/rfc/rfc9180.html
+pub static PROVIDER: &'static dyn HpkeProvider = &Provider {};
+
+/// Provider maps a supported [HpkeSuite] to a [SuiteContext], which can then be used to seal/open.
+#[derive(Debug)]
+struct Provider {}
+
+impl HpkeProvider for Provider {
+    fn start(&self, suite: &HpkeSuite) -> Result<Box<dyn Hpke + 'static>, Error> {
+        Ok(Box::new(SuiteContext::new(suite)?))
+    }
+
+    fn supports_suite(&self, suite: &HpkeSuite) -> bool {
+        SuiteContext::supports_suite(suite)
+    }
+}
+
+/// SuiteContext holds the concrete instantiations of the algorithms from the suite.
+struct SuiteContext {
+    suite: HpkeSuite,
+    dh_kem: Kem,
+    hkdf: &'static dyn Hkdf,
+    aead: &'static aead::Algorithm,
+}
+
+impl Debug for SuiteContext {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        // `&dyn Hkdf` is not Debug and the suite itself contains all the required context.
+        f.debug_struct("SuiteContext")
+            .field("suite", &self.suite)
+            .finish()
+    }
+}
+
+impl SuiteContext {
+    fn new(suite: &HpkeSuite) -> Result<Self, Error> {
+        Ok(Self {
+            suite: *suite,
+            dh_kem: Kem::new(&suite.kem)?,
+            hkdf: Self::hkdf_algorithm(suite)
+                .ok_or(Error::General("unsupported KDF algorithm".into()))?,
+            aead: Self::aead_algorithm(suite)
+                .ok_or(Error::General("unsupported AEAD algorithm".into()))?,
+        })
+    }
+
+    fn supports_suite(suite: &HpkeSuite) -> bool {
+        // an HPKE suite is supported if we support all of:
+        //   * the KEM algorithm
+        //   * the HKDF algorithm
+        //   * the AEAD algorithm
+        Kem::supports_suite(&suite.kem)
+            && Self::hkdf_algorithm(suite).is_some()
+            && Self::aead_algorithm(suite).is_some()
+    }
+
+    fn hkdf_algorithm(suite: &HpkeSuite) -> Option<&'static dyn Hkdf> {
+        Some(match suite.sym.kdf_id {
+            HpkeKdf::HKDF_SHA256 => HKDF_SHA256,
+            HpkeKdf::HKDF_SHA384 => HKDF_SHA384,
+            HpkeKdf::HKDF_SHA512 => HKDF_SHA512,
+            HpkeKdf::Unknown(_) => return None,
+        })
+    }
```

Comment:
Great stuff on getting this working!

Generally I like to make things like this "linker-friendly" in the sense that the linker can throw away unused algorithms. These sort of functions that map an enum to an algorithm implementation would be in the way of that.

The way that could look here is:

- have static `HpkeContext` values, one for each supported `HpkeSuite`, referring to `&aead::AES_128_GCM` etc.
- `Provider` is initialised with a list of these, it `supports_suite` if one exists in its list, and for `start()` can look up the algorithm details similarly, then pass that into the object that is `dyn Hpke`

Reply:
That makes sense 👍 I think this arrangement might also help me knock out a TODO I left RE: compile time constant output sizes in labelled expand.


I'll adjust.

Reply:
Reworked to be linker-friendly.